nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -35,338 +35,406 @@
-    tf2::durationFromSec(0.1), &tf_error))
-  {
-    if (last_error + nav2_util::duration_from_seconds(5.0) < now()) {
-      RCLCPP_WARN(
-        get_logger(),
-        "Timed out waiting for transform from %s to %s to become available before running costmap, tf error: %s", //NOLINT
-        robot_base_frame_.c_str(), global_frame_.c_str(), tf_error.c_str());
-      last_error = now();
-    }
-    // The error string will accumulate and errors will typically be the same, so the last
-    // will do for the warning above. Reset the string here to avoid accumulation.
-    tf_error.clear();
+Costmap2DROS::~Costmap2DROS()
+{
+  RCLCPP_INFO(get_logger(), "Destroying");
+}
+
+nav2_lifecycle::CallbackReturn
+Costmap2DROS::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+  getParameters();
+
+  // Create the costmap itself
+  layered_costmap_ = new LayeredCostmap(global_frame_, rolling_window_, track_unknown_space_);
+
+  if (!layered_costmap_->isSizeLocked()) {
+    layered_costmap_->resizeMap((unsigned int)(map_width_meters_ / resolution_),
+      (unsigned int)(map_height_meters_ / resolution_), resolution_, origin_x_, origin_y_);
   }
 
-  // check if we want a rolling window version of the costmap
-  bool rolling_window, track_unknown_space, always_send_full_costmap;
-  get_parameter_or<bool>("rolling_window", rolling_window, false);
-  get_parameter_or<bool>("track_unknown_space", track_unknown_space, false);
-  get_parameter_or<bool>("always_send_full_costmap", always_send_full_costmap, false);
+  // Create the transform-related objects
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(rclcpp_node_->get_clock());
+  tf_buffer_->setUsingDedicatedThread(true);
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_, rclcpp_node_, false);
 
-  layered_costmap_ = new LayeredCostmap(global_frame_, rolling_window, track_unknown_space);
+  // Then load and add the plug-ins to the costmap
+  for (int i = 0; i < plugin_names_.size(); ++i) {
+    RCLCPP_INFO(get_logger(), "Using plugin \"%s\"", plugin_names_[i].c_str());
 
-  if (plugin_names.size() == plugin_types.size()) {
-    for (int i = 0; i < plugin_names.size(); ++i) {
-      RCLCPP_INFO(get_logger(), "Using plugin \"%s\"", plugin_names[i].c_str());
-      std::shared_ptr<Layer> plugin = plugin_loader_.createSharedInstance(plugin_types[i]);
-      layered_costmap_->addPlugin(plugin);
-      plugin->initialize(layered_costmap_, plugin_names[i], &tf_, node_);
-    }
-  } else {
-    std::string plugin_error = "Plugin Name and Plugin Type sizes do not match";
-    RCLCPP_ERROR(get_logger(), plugin_error);
-    throw std::runtime_error(plugin_error);
+    std::shared_ptr<Layer> plugin = plugin_loader_.createSharedInstance(plugin_types_[i]);
+    layered_costmap_->addPlugin(plugin);
+
+    // TODO(mjeronimo): instead of get(), use a shared ptr
+    plugin->initialize(layered_costmap_, plugin_names_[i], tf_buffer_.get(),
```

Comment:
On the re-declaration of parameters point before, I think the plugin parameters should be ok since you are recreating the plugins from scratch here anyway. However, in the reset case (see other comment), we'd need to clean them up if the are re-declaring parameters